### ISSUE TITLE
 dbus broker errors "ignoring duplicate name"

### ISSUE THREAD

**Issue by kreutzi** [NixOS/nixpkgs#303078]
### Describe the bug
journalctl -x -b -p 3 output gives dbus-broker-launch ignoring duplicate name errors 
example error: Apr 10 12:14:13 nixos dbus-broker-launch[1434]: Ignoring duplicate name 'org.kde.kscreen.osdService' in service file '/nix/store/kn72wm> 

### Steps To Reproduce
1. switch to unstable channel
2. enable flake
3. upgrading

### Expected behavior
no dbus-broker errors 

### Screenshots
![image](https://github.com/NixOS/nixpkgs/assets/88722623/34df00d8-017c-4106-ba9b-c52b2082ca3c)

### Notify maintainers
@LDprg 

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
- system: `"x86_64-linux"`
 - host os: `Linux 6.6.25, NixOS, 24.05 (Uakari), 24.05.20240408.4cba8b5`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/450afzqlzzgw6wnyc3dwysf3i5yxyqkr-source`

```

---

**Comment by Aleksanaa** [comment]
related: https://github.com/NixOS/nixpkgs/pull/299812

**Comment by Aleksanaa** [comment]
Have you tried rebooting to see if the error persists?

**Comment by LDprg** [comment]
I don't think those errors are too much of a concern.
Seems just like the error says duplicated names in the service files, which is more of a warning than an error.

You might open one of the service files and look if there are actually duplicated names inside.

**Comment by kreutzi** [comment]
> I don't think those errors are too much of a concern. Seems just like the error says duplicated names in the service files, which is more of a warning than an error.
> 
> You might open one of the service files and look if there are actually duplicated names inside.

Apr 10 13:03:34 nixos dbus-broker-launch[1445]: Ignoring duplicate name 'org.freedesktop.systemd1' in service file '/nix/store/kn72wm4pw2997yxpdvys2z1hvs4msa7p-system-path/share/dbus-1/services/org.freedesktop.systemd1.service'

file content 

#  SPDX-License-Identifier: LGPL-2.1-or-later
#
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.

[D-BUS Service]
Name=org.freedesktop.systemd1
Exec=/nix/store/ifzwv2xqwdnv1gz87rxkizi67py5p3vj-coreutils-9.4/bin/false
User=root

**Comment by LDprg** [comment]
@peterhoeg maybe you know what's going on? It seems weird to me since there is no obvious reason why the error appears.

**Comment by LDprg** [comment]
Also maybe this is related https://github.com/bus1/dbus-broker/issues/339

**Comment by kreutzi** [comment]
Its weird doesn't happen on a fresh install though

**Comment by kreutzi** [comment]
@LDprg i investigated more into this and i keep finding weird errors 
this is a portion of the output for journalctl -u dbus-broker
having invalid username for pulse and no entry for nm-openconnect and invalid configs
```
- Boot 515786cc9ee846049e2a3413f71989e1 --
Apr 10 21:42:33 nixos systemd[1]: Starting D-Bus System Message Bus...
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Looking up NSS user entry for 'nm-openconnect'...
Apr 10 21:42:33 nixos dbus-broker-launch[954]: NSS returned no entry for 'nm-openconnect'
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Invalid user-name in /nix/store/khyv34c7rdrw34cfkq44cqlly6gfja3v-system-path/share/dbus-1/system.d/nm-openconnect-service.conf +9: user="nm-openconnect"
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Looking up NSS user entry for 'pulse'...
Apr 10 21:42:33 nixos dbus-broker-launch[954]: NSS returned no entry for 'pulse'
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Invalid user-name in /nix/store/khyv34c7rdrw34cfkq44cqlly6gfja3v-system-path/share/dbus-1/system.d/pulseaudio-system.conf +27: user="pulse"
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Invalid user-name in /nix/store/c469rx36lmx81mgpnhc5wa23pmmr8wcx-NetworkManager-openconnect-1.2.10/share/dbus-1/system.d/nm-openconnect-service.conf +9: user="nm-openconnect"
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'org.opensuse.CupsPkHelper.Mechanism' in service file '/nix/store/c8s741qsm277h12zmln6qdrcaqf24y1d-cups-pk-helper-0.2.6/share/dbus-1/system-services/org.opensuse.Cu>
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'org.freedesktop.ModemManager1' in service file '/nix/store/9xjzisxj6qzz537bjw9i5m923cnbvmha-modemmanager-1.22.0/share/dbus-1/system-services/org.freedesktop.ModemM>
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'org.freedesktop.nm_dispatcher' in service file '/nix/store/p7m9nbg1f13kqvafahpl3q2fbfa3j6hw-networkmanager-1.46.0/share/dbus-1/system-services/org.freedesktop.nm_d>
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'org.freedesktop.nm_priv_helper' in service file '/nix/store/p7m9nbg1f13kqvafahpl3q2fbfa3j6hw-networkmanager-1.46.0/share/dbus-1/system-services/org.freedesktop.nm_>
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'fi.w1.wpa_supplicant1' in service file '/nix/store/c0cv1yk1gmqwf89sz3y5cahw5v5h6xhz-wpa_supplicant-2.10/share/dbus-1/system-services/fi.w1.wpa_supplicant1.service'
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'org.freedesktop.UPower' in service file '/nix/store/2lrhi78qfg2036mivcwzy91g1l85jyab-upower-1.90.2/share/dbus-1/system-services/org.freedesktop.UPower.service'
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'org.freedesktop.UDisks2' in service file '/nix/store/p1kcwvyxwhsb6cl9vz1npgwmaf9l99rq-udisks-2.10.1/share/dbus-1/system-services/org.freedesktop.UDisks2.service'
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'net.hadess.PowerProfiles' in service file '/nix/store/zgzhcpywdy7rfjwi97xrv4mismn2cqsc-power-profiles-daemon-0.20/share/dbus-1/system-services/net.hadess.PowerProf>
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'org.freedesktop.UPower.PowerProfiles' in service file '/nix/store/zgzhcpywdy7rfjwi97xrv4mismn2cqsc-power-profiles-daemon-0.20/share/dbus-1/system-services/org.free>
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'org.bluez.mesh' in service file '/nix/store/468hykzb473ascad7pkcdfi6sg2i73kz-bluez-5.72/share/dbus-1/system-services/org.bluez.mesh.service'
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'org.bluez' in service file '/nix/store/468hykzb473ascad7pkcdfi6sg2i73kz-bluez-5.72/share/dbus-1/system-services/org.bluez.service'
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'org.freedesktop.Accounts' in service file '/nix/store/371pd4m7h5y93687ifz4mvh3rkb3x72a-accountsservice-23.13.9/share/dbus-1/system-services/org.freedesktop.Account>
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'org.freedesktop.RealtimeKit1' in service file '/nix/store/w55j5y4wa9dk1sqwh0vziadb0inyxqh2-rtkit-0.13/share/dbus-1/system-services/org.freedesktop.RealtimeKit1.ser>
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ignoring duplicate name 'org.freedesktop.PolicyKit1' in service file '/nix/store/zrvqzsn2q0101i395igwcxqk9413r0dh-polkit-123/share/dbus-1/system-services/org.freedesktop.PolicyKit1.service'
Apr 10 21:42:33 nixos systemd[1]: Started D-Bus System Message Bus.
Apr 10 21:42:33 nixos dbus-broker-launch[954]: Ready
Apr 10 21:45:58 nixos dbus-broker-launch[954]: Activation request for 'org.freedesktop.resolve1' failed: The systemd unit 'dbus-org.freedesktop.resolve1.service' could not be found.

```

**Comment by LDprg** [comment]
@kreutzi did you only experience the errors or are there any other problems (like package not working as intended)?

**Comment by kreutzi** [comment]
> @kreutzi did you only experience the errors or are there any other problems (like package not working as intended)?

haven't used this system for that long only 2 days but everything is working i will keep you updated

**Comment by kreutzi** [comment]
@LDprg @Aleksanaa 
got some graphical bugs with kde randomly invalidating program executables then i investigated some more but pinpointed at kde 

found out two dbus services with the same exec but different names 

org.kde.kwalletd5
[D-BUS Service]
Name=org.kde.kwalletd5
Exec=/nix/store/18n78qdjgylsz65jv8dp250y6hjcvbp6-kwallet-6.0.0/bin/kwalletd6

org.kde.kwalletd6
[D-BUS Service]
Name=org.kde.kwalletd6
Exec=/nix/store/18n78qdjgylsz65jv8dp250y6hjcvbp6-kwallet-6.0.0/bin/kwalletd6

i would guess there are similar examples to this but should this cause the errors i am seeing ?

**Comment by Aleksanaa** [comment]
I think the reason for this is that the kde6 modules are not handled correctly, and some packages from kde5 are mixed in it

**Comment by kreutzi** [comment]
> I think the reason for this is that the kde6 modules are not handled correctly, and some packages from kde5 are mixed in it


Yeah i switched desktops to test and the errors lessened to about 30 errors only about some remaining org.freedesktop packages

**Comment by LDprg** [comment]
@K900 can you take a look at this seems like an error with kde 6 packages. Likely some old kde 5 packages get included along with the kde 6 ones.

@kreutzi The freedesktop ones are very likely just stuff like multiple file explorers that are installed on the system.

**Comment by SuperSandro2000** [comment]
Yeah, those are just warnings. The old dbus just silently ignored it. Has anyone ask idea how we can work around this?

**Comment by peterhoeg** [comment]
> I think the reason for this is that the kde6 modules are not handled correctly, and some packages from kde5 are mixed in it

This issue predates the inclusion of plasma6. It's a warning as others have pointed out but it really should be fixed.

**Comment by kreutzi** [comment]
> Yeah, those are just warnings. The old dbus just silently ignored it. Has anyone ask idea how we can work around this?

I came to the same conclusion dbus-broker seems more verbose than the old dbus-daemon.  

As @peterhoeg mentioned: 
> This issue predates the inclusion of plasma6

This is beyond my expertise, even after extensive debugging.  Perhaps we could involve some KDE developers?

**Comment by kreutzi** [comment]
This is an explanation for said errors 

Subject: Invalid service file
░░ Defined-By: dbus-broker
░░ Support: https://groups.google.com/forum/#!forum/bus1-devel
░░ 
░░ A service file is a ini-type configuration file.
░░ 
░░ It has one required section
░░ named [D-BUS Service]. The section contains the required key 'Name', which
░░ must be a valid D-Bus name that is unique across all service files. It also
░░ contains at least one of the two optional keys 'SystemdService' and 'Exec',
░░ as well as optionally the key 'User'. Exec must be a valid shell command and
░░ User must be a valid user on the system.
░░ 
░░ A service file should be named after the D-Bus name it configures. That is
░░ a file containing Name=org.foo.bar1 should be named org.foo.bar1.service.
░░ For backwards compatibility, we only warn when files do not follow this
░░ convention when run as a user bus. The system bus considers this an error
░░ and ignores the service file.

**Comment by eclairevoyant** [comment]
The straightforward solution is to fix the offending packages, one-by-one.
I've opened an example PR to discuss further - #373282

**Comment by marienz** [comment]
At least on my system, I think most of these are there because dbus-broker is loading the same .service file twice through different paths/symlinks.

If I make the dbus-broker logging a little more verbose, including the path it loaded the conflicting service definition from:

```patch
diff --git a/src/launch/launcher.c b/src/launch/launcher.c
index 5985712..f189f4a 100644
--- a/src/launch/launcher.c
+++ b/src/launch/launcher.c
@@ -723,7 +723,7 @@ static int launcher_load_service_file(Launcher *launcher, const char *path, cons
                         log_append_service_path(&launcher->log, path);
                         log_append_service_name(&launcher->log, name);

-                        r = log_commitf(&launcher->log, "Ignoring duplicate name '%s' in service file '%s'\n", name, path);
+                        r = log_commitf(&launcher->log, "Ignoring duplicate name '%s' in service file '%s' (already seen in '%s')\n", name, path, old_service->data->path);
                         if (r)
                                 return error_fold(r);

```

it complains about paths like:

```
Jul 12 14:26:30 jian dbus-broker-launch[974]: Ignoring duplicate name 'org.freedesktop.UPower' in service file '/nix/store/50g7m9svx09bhllrfdgy1j4khwj2pn5b-upower-1.90.6/share/dbus-1/system-services/org.freedesktop.UPower.service' (already seen in '/nix/store/xz0r0ndpgivqi17gxcvmkz1a0yvryjl6-system-path/share/dbus-1/system-services/org.freedesktop.UPower.service')
```

Those are the same file by different names: the system-path one is a symlink to the upower one.

I *think* this happens because /etc/dbus-1/system.conf contains `<servicedir>` entries for both /nix/store/xz0r0ndpgivqi17gxcvmkz1a0yvryjl6-system-path/share/dbus-1/system-services and individual packages with symlinks to their .service files in system-path.

Similarly:

```
Jul 12 14:26:32 jian dbus-broker-launch[1565]: Ignoring duplicate name 'org.freedesktop.impl.portal.PermissionStore' in service file '/run/current-system/sw/share/dbus-1/services/org.freedesktop.impl.portal.PermissionStore.service' (already seen in '/etc/profiles/per-user/marienz/share/dbus-1/services/org.freedesktop.impl.portal.PermissionStore.service')
```

These are again the same file, but for this one I haven't tracked down why it's looking for service files in /etc/profiles/per-user/marienz/share/dbus-1/services.

For these paths, "Ignoring duplicate" is entirely cosmetic, but especially for that last one (where one entry is provided by NixOS while the other is from home-manager) I suspect hiding the duplicates from dbus-broker is not very practical... Might make more sense to have dbus-broker canonicalize the paths to the service files and not emit that warning (or at least downgrade it to an informational message) if it's skipping a file whose canonicalized path is the same as the already-loaded duplicate.

(That way you'd still get the warning if the .service files in /run/current-system and /etc/profiles/per-user/ are actually different: that seems like a case still worth warning about.)

**Comment by abkein** [comment]
## Just for summary

The messages `dbus-broker-launch: Ignoring duplicate name` are caused by duplicate D-Bus activation service files, not by failing services.

Root cause:

- The evaluated NixOS `services.dbus.packages` contains both `system-path` and the individual packages inside it, for example `networkmanager`, `iwd`, `bluez`, `upower`, `udisks`, `polkit`, `xdg-desktop-portal`, `thunar`, `gvfs`, etc.
- `/etc/dbus-1/system.conf` and `/etc/dbus-1/session.conf` therefore expose the same `.service` files through two paths:
  - merged profile path such as `/nix/store/...-system-path/share/dbus-1/...`
  - original package path such as `/nix/store/...-networkmanager-.../share/dbus-1/...`
- `dbus-broker-launch` sees the same `Name=` twice and logs `Ignoring duplicate name ...` for the later copy.

So these are mostly noise from Nix/NixOS profile composition. `dbus-broker` keeps one activation entry and ignores the duplicate.

## Quick filter — doesn't work, see https://github.com/NixOS/nixpkgs/issues/303078#issuecomment-4698342707

```nix
{
  systemd.services.dbus-broker.serviceConfig.LogFilterPatterns = [
    "~Ignoring duplicate name '.*' in service file '.*'"
  ];

  systemd.user.services.dbus-broker.serviceConfig.LogFilterPatterns = [
    "~Ignoring duplicate name '.*' in service file '.*'"
  ];
}
```

**Comment by marienz** [comment]
[dbus-broker-logging.patch](https://github.com/user-attachments/files/28911250/dbus-broker-logging.patch) applied with
```nix
    services.dbus = {
      brokerPackage = pkgs.dbus-broker.overrideAttrs (old: {
        patches =
          (old.patches or [ ])
          ++ [
            ./dbus-broker-logging.patch
          ];
      });
    };
```

should filter out just the actual duplicates: it skips the warning if the canonicalized paths are the same. (I've had this applied for a while, kind of forgot about it)

(Haven't tried upstreaming this: it seems like a pretty niche problem outside NixOS...)

**Comment by fidgetingbits** [comment]
> ## Quick workaround (filtering)
> 
> {
>   systemd.services.dbus-broker.serviceConfig.LogFilterPatterns = [
>     "~Ignoring duplicate name '.*' in service file '.*'"
>   ];
> 
>   systemd.user.services.dbus-broker.serviceConfig.LogFilterPatterns = [
>     "~Ignoring duplicate name '.*' in service file '.*'"
>   ];
> }

Did you actually get that to work? I notice it's commented out in your public nix config, and didn't work for me (not sure why). According to the [this](https://github.com/systemd/systemd/issues/31113) `LogFilterPatterns` isn't supported by user services, so the second part definitely won't work iiuc.

**Comment by abkein** [comment]
> Did you actually get that to work? I notice it's commented out in your public nix config, and didn't work for me (not sure why). According to the [this](https://github.com/systemd/systemd/issues/31113) `LogFilterPatterns` isn't supported by user services, so the second part definitely won't work iiuc.

@fidgetingbits, you're absolutely right, it doesn't work. It seems after I applied it I simply didn't check if it works, so sorry for misleading info. However, by design it had to.
Nonetheless, it was a workaround and the solution in https://github.com/NixOS/nixpkgs/pull/549241 seems to be a proper one, if we still want `services.dbus.packages` to include `system.path`, while also adding packages to it in modules.
However, I personally don't like the concept of `system.path` as a "garbage" collector and think it'd be better if packages themselves were managing what to put in, e.g. `/share`, add to `$PATH` and DBus, etc. But this would require a major changes in package API, so I don't even dream of that.

**Comment by marienz** [comment]
> the solution in https://github.com/NixOS/nixpkgs/pull/549241 seems to be a proper one, if we still want services.dbus.packages to include system.path, while also adding packages to it in modules.

I haven't looked closely, but I think that'll fix cases like the first one mentioned in https://github.com/NixOS/nixpkgs/issues/303078#issuecomment-3064686826 (with `/nix/store/xz0r0ndpgivqi17gxcvmkz1a0yvryjl6-system-path/share/dbus-1/system-services/org.freedesktop.UPower.service` versus `/nix/store/50g7m9svx09bhllrfdgy1j4khwj2pn5b-upower-1.90.6/share/dbus-1/system-services/org.freedesktop.UPower.service`) but can't fix the second one (with `/etc/profiles/per-user/marienz/share/dbus-1/services/org.freedesktop.impl.portal.PermissionStore.service` versus `/run/current-system/sw/share/dbus-1/services/org.freedesktop.impl.portal.PermissionStore.service`).

(That first one looks like the scenario you describe that PR as fixing (one from `system.path`, the other from a module adding to `services.dbus.packages`). But the second one is user and system profile colliding, which doesn't seem fixable at build time...)

(The PR might still be a good idea just to cut down on noise, of course.)
